storage/authorization.go

```diff
@@ -0,0 +1,220 @@
+// Package storage provides clients for Microsoft Azure Storage Services.
+package storage
+
+import (
+	"fmt"
+	"net/url"
+	"sort"
+	"strings"
+)
+
+// See: https://docs.microsoft.com/rest/api/storageservices/fileservices/authentication-for-the-azure-storage-services
+
+type authentication string
+
+const (
+	sharedKey             authentication = "sharedKey"
+	sharedKeyForTable     authentication = "sharedKeyTable"
+	sharedKeyLite         authentication = "sharedKeyLite"
+	sharedKeyLiteForTable authentication = "sharedKeyLiteTable"
+
+	// headers
+	headerAuthorization     = "Authorization"
+	headerContentLength     = "Content-Length"
+	headerDate              = "Date"
+	headerXmsDate           = "x-ms-date"
+	headerXmsVersion        = "x-ms-version"
+	headerContentEncoding   = "Content-Encoding"
+	headerContentLanguage   = "Content-Language"
+	headerContentType       = "Content-Type"
+	headerContentMD5        = "Content-MD5"
+	headerIfModifiedSince   = "If-Modified-Since"
+	headerIfMatch           = "If-Match"
+	headerIfNoneMatch       = "If-None-Match"
+	headerIfUnmodifiedSince = "If-Unmodified-Since"
+	headerRange             = "Range"
+)
+
+func (c Client) addAuthorizationHeader(verb, url string, headers *map[string]string, auth authentication) error {
+	authHeader, err := c.getSharedKey(verb, url, *headers, auth)
+	if err != nil {
+		return err
+	}
+	(*headers)[headerAuthorization] = authHeader
+	return nil
+}
+
+func (c Client) getSharedKey(verb, url string, headers map[string]string, auth authentication) (string, error) {
+	canRes, err := c.buildCanonicalizedResource(url, auth)
+	if err != nil {
+		return "", err
+	}
+
+	canString := c.buildCanonicalizedString(verb, headers, canRes, auth)
+	return c.createAuthorizationHeader(canString, auth), nil
+}
+
+func (c Client) buildCanonicalizedResource(uri string, auth authentication) (string, error) {
```

Comment:
Like below, this method would achieve better performance and not make as many allocations if it was rewritten to use buffers like .NET StringBuilder instead of just using string concatenation.